### PR TITLE
Remove private DNS zones from architecture

### PR DESCRIPTION
**Pull request by jim-minter**
Our/OCP's use of a private DNS zone prevents customers from overriding the vnet DNS.  This PR removes private DNS zones from our architecture.

Current status:
* feedback welcome
* deployment manually tested in default/non-default domain, public/private API server and public/private console contexts
* upgrade of existing clusters implemented and tested
* unit tests are a WIP

Design ideas:
1. at node level, interpose dnsmasq between clients and the vnet resolver; inject additional domains as necessary.  This should cover off all node-level and pod-level DNS requests.  I quite like this concept as we could later build on it for egress lockdown work.  The main difficulty is populating the dnsmasq configuration inside the MachineConfig, hence:
2. preallocate api, api-int and default router IPs much earlier in our cluster creation process so that we can lay down the above dnsmasq configuration in the install graph.  The main wrinkle here is that there is a race condition introduced when the default router has private visibility, but I think we can probably get away with it.  There is also upside here: by predefining these IPs up front, I think we could probably later refactor/simplify/speeden our post-install processes (e.g. pre-configure certificates and maybe avoid the associated API server rotation?)

Implementation summary:
1. [Installer changes](https://github.com/jim-minter/installer/compare/release-4.6...release-4.6-azure-lbip)
   * add the `bootkube.ARODNSConfig` type to pass IPs into the installer (a bit clunky)
   * populate the openshift-ingress/router-default LoadBalancer Service early, configuring it to use the IP we've already allocated.  Wrinkle: currently the ingress operator fails to adopt the Service later; I'm not sure if that matters or not
   * configure and enable dnsmasq on bootstrap
   * configure and enable dnsmasq on masters and workers via MachineConfigs
   * modify machine config server URL and certificate to use api-int IP address as well as DNS, as the DNS not resolvable at the point that non-bootstrap ignition calls the API
   * disable ingress operator dynamic DNS (prevent trying to write to the private DNS zone which no longer exists).
2. This PR:
   * create network resources (public IP(s), LBs, PLS for consistency) before install graph creation, so we can discover their IPs
   * if the default router has public visibility, create its public IP up front; it will be [adopted](https://docs.microsoft.com/en-us/azure/aks/static-ip#create-a-service-using-the-static-ip-address) later by the early populated openshift-ingress/router-default LoadBalancer Service
   * if the default router has private visibility, discover the highest unused subnet IP address; it will be adopted later by the early populated openshift-ingress/router-default LoadBalancer Service.  We hope that no-one else takes the IP address in the interim
   * add api-int IP to internal model (I'm not actually sure this is necessary or desirable)
   * pass discovered IPs to installer
   * remove dynamic vnet dns validation and upgrade geneva action (this may be premature, depending on the upgrade story)
   * remove deployment of private dns zones (but not deletion, at least for onw)
   * implement upgrade

Test plan:
- [x] create new-style cluster
- [x] admin upgrade new-style cluster (no-op)
- [x] validate new controller behaviour (modify/delete configs and inspect)
- [x] create and upgrade old-style cluster
- [x] scale up upgraded cluster
- [x] redeploy an old cluster node on an upgrade cluster
- [x] openshift upgrade upgrade cluster
- [x] create new-style cluster with non-standard dns server, non-standard cluster domain
- [x] validate node hostnames
- [x] openshift upgrade upgrade cluster
- [x] create new-style cluster with non-standard dns server, non-standard cluster domain, private API server and ingress
- [x] openshift upgrade upgrade cluster

### CODE REVIEW

**Comment by mjudeikis** [comment]
@jim-minter I think this is bit over the line how much we do to make OCP ARO. 
I think it would be good to get upstream alignment on how this gonna play in the long term OCP plans. This can easily backfire for ARO in the long term if we implement something which will contradict wider product development. 

If overall product direction will be similar (dnsmasq for the start) otherwise this sounds like 3.11 all over again.

File: pkg/frontend/adminactions/kubeactions.go [hunk 1]
```diff
@@ -31,14 +30,10 @@ type kubeActions struct {
 	oc        *api.OpenShiftCluster
 	dh        dynamichelper.Interface
 	configcli configclient.Interface
-
-	virtualNetworks network.VirtualNetworksClient
 }
 
 // NewKubeActions returns a kubeActions
-func NewKubeActions(log *logrus.Entry, env env.Interface, oc *api.OpenShiftCluster,
-	subscriptionDoc *api.SubscriptionDocument) (KubeActions, error) {
```
Comment (nilsanderselde) [inline pkg/frontend/adminactions/kubeactions.go:40]: pretty cool that this PR also cleans up KubeActions a bit

**Comment by cgwalters** [comment]
This seems strongly related to https://github.com/openshift/machine-config-operator/pull/2258

File: pkg/cluster/fixmcsuserdata.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	if machineProviderSpec.UserDataSecret.Namespace == "" {
+		machineProviderSpec.UserDataSecret.Namespace = objMeta.Namespace
+	}
+
+	return machineProviderSpec.UserDataSecret, nil
+}
+
+func (m *manager) fixMCSUserData(ctx context.Context) error {
```
Comment (mjudeikis) [inline pkg/cluster/fixmcsuserdata.go:109]: Can you elaborate why we need this fixup?

File: pkg/cluster/samples.go [hunk 3]
```diff
@@ -61,14 +61,23 @@ func (m *manager) disableOperatorHubSources(ctx context.Context) error {
 				Name:     "certified-operators",
 				Disabled: true,
 			},
+			{
```
Comment (mjudeikis) [inline pkg/cluster/samples.go:64]: Does not look related to this PR

File: pkg/operator/controllers/dnsmasq/cluster_controller.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	"github.com/Azure/ARO-RP/pkg/util/dynamichelper"
+)
+
+type ClusterReconciler struct {
```
Comment (mjudeikis) [inline pkg/operator/controllers/dnsmasq/cluster_controller.go:23]: Readme what this does and why would be great

File: pkg/operator/controllers/dnsmasq/cluster_controller.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+		r.log.Error(err)
+		return reconcile.Result{}, err
+	}
+
+	roles := make([]string, 0, len(mcps.Items))
+	for _, mcp := range mcps.Items {
+		roles = append(roles, mcp.Name)
+	}
+
+	err = reconcileRoles(ctx, r.arocli, r.dh, roles...)
```
Comment (mjudeikis) [inline pkg/operator/controllers/dnsmasq/cluster_controller.go:60]: `reconcileRole` name is very confusing. What roles this is? Kubernetes roles? Node roles? Azure roles? Took me a while to read down the rabbit hole to understand.

**Review by mjudeikis — CHANGES_REQUESTED** [review]
This is something :) 
I think ok with most of the changes except few cosmetics changes.

**Comment by mjudeikis** [comment]
Few other questions:
1. At what point we added ARO worker registry :
https://github.com/jim-minter/installer/commit/6fb6f85c826e6f2b15e80c113c9f782abe4361e6
2. From installer, &bootkube.ARODNSConfig{} worth it making more general? AROConfig? For future re-usability and try to upstream it ?
3. loggingConfig, aroDNSConfig - could be done by (2) and simplified?
4. Not clear why existing behavior for non ARO changes:
https://github.com/jim-minter/installer/commit/9f4cb3411dd1002c947f4b926f1f4d6377954145#diff-2d86052ba31adcc485733d72fc5da379a06d23af4b276d92b472b0e5a5461b5cR52

File: pkg/cluster/samples.go (hunk 3 above)
Comment (jim-minter) [reply pkg/cluster/samples.go:64]: Correct; the first few commits in this PR are like that.

File: pkg/operator/controllers/dnsmasq/cluster_controller.go (hunk 5 above)
Comment (jim-minter) [reply pkg/operator/controllers/dnsmasq/cluster_controller.go:60]: Agree, will fix and document

**Comment by jim-minter** [comment]
> Few other questions:
> 
> 1. At what point we added ARO worker registry :
> jim-minter/installer@6fb6f85

https://github.com/Azure/ARO-RP/pull/1167

> 2. From installer, &bootkube.ARODNSConfig{} worth it making more general? AROConfig? For future re-usability and try to upstream it ?

Yes, there's space to modify this in the future.

> 3. loggingConfig, aroDNSConfig - could be done by (2) and simplified?

Feasibly, yes.

> 4. Not clear why existing behavior for non ARO changes:
> jim-minter/installer@9f4cb34#diff-2d86052ba31adcc485733d72fc5da379a06d23af4b276d92b472b0e5a5461b5cR52

It does not.  It is a copy of the `default:` code which otherwise would not be reached.

**Review by mjudeikis — DISMISSED** [review]
